modified version of SCISSOR's `build_gaf` function

LedaKatopodi opened the pull request:
SCISSOR's `build_gaf` function seems to have been modified, and the `Gene` argument seems to have been removed. It also browses through the whole GTF file (input) and exports the regions (the exons per gene, in a concatenated string form) for all genes included in the annotation file. This significantly increases runtime, whereas no option to select for genes of interest is given.

In this modified version:

1. A list with genes of interest can be supplied as argument. The function subsets the input file, and returns only the regions (exons) for the selected genes. If no gene list is specified, the function operates on the initial input file and will return the regions for all genes present in the annotation.

2. The function prioritizes a single annotation source for which to return the regions. Annotation priority is as follows: **Ensembl-Havana > Ensembl > Havana**. This was done in hope of alleviating issues similar to those described in [issue #3](https://github.com/hyochoi/SCISSOR/issues/3), which seem to stem from the concatenation of exons from all available sources. The annotation prioritization was based on the fact that *"for human and mouse, this combined Ensembl/HAVANA gene set is the default gene set from the GENCODE project"* ([source](https://useast.ensembl.org/info/genome/genebuild/annotation_merge.html)).

**DISCLAIMER**: The provided code was tested only for a small number of genes (6) on mouse (Mus musculus, GRCm38, Ensembl 98); R version = 4.0.1.

If you find any bugs, feel free to open a ticket.

Cheers,
Leda

LedaKatopodi committed:
modified version of SCISSOR's `build_gaf` function

SCISSOR's `build_gaf` function seems to have been modified, and the `Gene` argument seems to have been removed. It also browses through the whole GTF file (input) and exports the regions (the exons per gene, in a concatenated string form) for all genes included in the annotation file. This significantly increases runtime, whereas no option to select for genes of interest is given.

In this modified version:

1. A list with genes of interest can be supplied as argument. The function subsets the input file, and returns only the regions (exons) for the selected genes. If no gene list is specified, the function operates on the initial input file and will return the regions for all genes present in the annotation.

2. The function prioritizes a single annotation source for which to return the regions. Annotation priority is as follows: **Ensembl-Havana > Ensembl > Havana**. This was done in hope of alleviating issues similar to those described in [issue #3](https://github.com/hyochoi/SCISSOR/issues/3), which seem to stem from the concatenation of exons from all available sources. The annotation prioritization was based on the fact that *"for human and mouse, this combined Ensembl/HAVANA gene set is the default gene set from the GENCODE project"* ([source](https://useast.ensembl.org/info/genome/genebuild/annotation_merge.html)).

**DISCLAIMER**: The provided code was tested only for a small number of genes (6) on mouse (Mus musculus, GRCm38, Ensembl 98); R version = 4.0.1.

If you find any bugs, feel free to open a ticket.

Cheers,
Leda

canankolakoglu commented:
Hi,
I've try to run the modified version of `build_gaf` function which is `BuildExonsFromGTF_GeneSelection`. 
Here the code:
`BuildExonsFromGTF_GeneSelection(geneList="HBA1", GTFfile="/gencode.v24.primary_assembly.annotation.gtf")`
When I run this code, I get this error:

> Error in read.table(gffFile, sep = "\t", as.is = TRUE, quote = "", header = FALSE, : no lines available in input

> 4. stop("no lines available in input")
> 3. read.table(gffFile, sep = "\t", as.is = TRUE, quote = "", header = FALSE, comment.char = "#", nrows = nrows, colClasses = c("character", "character", "character", "integer", "integer", "character", "character", "character", "character"))
> 2. gffRead(GTFfile_temp)
> 1. BuildExonsFromGTF_GeneSelection(geneList = "HBA1", GTFfile = "/gencode.v24.primary_assembly.annotation.gtf")

I've checked `gffRead` function of `ballgown package` with my .gtf file and I was able to read the .gtf. 
So what could be the problem?

LedaKatopodi commented:
@canankolakoglu hi, 

In your GTF, are the gene symbols written in capital letters? Could you paste a couple of lines for the HBA1 gene here?

> Hi, I've try to run the modified version of `build_gaf` function which is `BuildExonsFromGTF_GeneSelection`. Here the code: `BuildExonsFromGTF_GeneSelection(geneList="HBA1", GTFfile="/gencode.v24.primary_assembly.annotation.gtf")` When I run this code, I get this error:
> 
> > Error in read.table(gffFile, sep = "\t", as.is = TRUE, quote = "", header = FALSE, : no lines available in input
> 
> > 4. stop("no lines available in input")
> > 5. read.table(gffFile, sep = "\t", as.is = TRUE, quote = "", header = FALSE, comment.char = "#", nrows = nrows, colClasses = c("character", "character", "character", "integer", "integer", "character", "character", "character", "character"))
> > 6. gffRead(GTFfile_temp)
> > 7. BuildExonsFromGTF_GeneSelection(geneList = smn1, GTFfile = "/gencode.v24.primary_assembly.annotation.gtf")
> 
> I've checked `gffRead` function of `ballgown package` with my .gtf file and I was able to read the .gtf. So what could be the problem?

canankolakoglu commented:
@LedaKatopodi Hi again,

Here is the table of gtf file that has been read by gffFile function;

![gffFiel](https://user-images.githubusercontent.com/32098605/147692393-7b7de1ea-2f78-48ec-b03b-de0e9b1e0354.png)

And here some lines for "HBA1" gene in text format;

> chr16	HAVANA	gene	176680	177522	.	+	.	gene_id "ENSG00000206172.8"; gene_type "protein_coding"; gene_status "KNOWN"; gene_name "HBA1"; level 2; havana_gene "OTTHUMG00000060138.3";

> chr16	HAVANA	transcript	176680	177522	.	+	.	gene_id "ENSG00000206172.8"; transcript_id "ENST00000320868.9"; gene_type "protein_coding"; gene_status "KNOWN"; gene_name "HBA1"; transcript_type "protein_coding"; transcript_status "KNOWN"; transcript_name "HBA1-001"; level 2; protein_id "ENSP00000322421.5"; tag "basic"; transcript_support_level "1"; tag "appris_principal_1"; tag "CCDS"; ccdsid "CCDS10399.1"; havana_gene "OTTHUMG00000060138.3"; havana_transcript "OTTHUMT00000133459.1";

> chr16	HAVANA	exon	176680	176811	.	+	.	gene_id "ENSG00000206172.8"; transcript_id "ENST00000320868.9"; gene_type "protein_coding"; gene_status "KNOWN"; gene_name "HBA1"; transcript_type "protein_coding"; transcript_status "KNOWN"; transcript_name "HBA1-001"; exon_number 1; exon_id "ENSE00001723291.1"; level 2; protein_id "ENSP00000322421.5"; tag "basic"; transcript_support_level "1"; tag "appris_principal_1"; tag "CCDS"; ccdsid "CCDS10399.1"; havana_gene "OTTHUMG00000060138.3"; havana_transcript "OTTHUMT00000133459.1";

> chr16	HAVANA	CDS	176717	176811	.	+	0	gene_id "ENSG00000206172.8"; transcript_id "ENST00000320868.9"; gene_type "protein_coding"; gene_status "KNOWN"; gene_name "HBA1"; transcript_type "protein_coding"; transcript_status "KNOWN"; transcript_name "HBA1-001"; exon_number 1; exon_id "ENSE00001723291.1"; level 2; protein_id "ENSP00000322421.5"; tag "basic"; transcript_support_level "1"; tag "appris_principal_1"; tag "CCDS"; ccdsid "CCDS10399.1"; havana_gene "OTTHUMG00000060138.3"; havana_transcript "OTTHUMT00000133459.1";

LedaKatopodi committed:
Updated `BuildExonsFromGTF_GeneSelection.R`

Modifications, based on the bugs reported [here](https://github.com/hyochoi/SCISSOR/pull/4#issuecomment-1002698179)

LedaKatopodi commented:
hi @canankolakoglu, thank you for your reply. 

I updated the code taking into account your input. Let me know if there are any additional issues.

Cheers,
Leda

LedaKatopodi committed:
Update BuildExonsFromGTF_GeneSelection.R

LedaKatopodi committed:
Update BuildExonsFromGTF_GeneSelection.R
